mache-vertx/src/main/java/com/excelian/mache/vertx/MacheInstanceCache.java

```diff
@@ -0,0 +1,85 @@
+package com.excelian.mache.vertx;
+
+import com.excelian.mache.core.Mache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Supplier;
+
+/**
+ * Manages the Mache instances created by the vertx REST endpoint
+ * <p>
+ * The methods exposed from this class are threadsafe.
+ */
+public class MacheInstanceCache {
+    private static final Logger LOG = LoggerFactory.getLogger(MacheInstanceCache.class);
+    private final ConcurrentMap<String, Mache<String, String>> cacheInstances = new ConcurrentHashMap<>();
+    private final Supplier<Mache<String, String>> factory;
+
+    /**
+     * Creates an instance cache that will retain the queried instances.
+     *
+     * @param factory A threadsafe factory to dispense new map instances
+     */
+    public MacheInstanceCache(Supplier<Mache<String, String>> factory) {
+        this.factory = factory;
+    }
+
+    private synchronized Mache<String, String> createMap(String mapId) {
+        LOG.trace("adding map {} to cache", mapId);
+        Mache<String, String> newMache = null;
+        try {
+            newMache = factory.get();
+            cacheInstances.put(mapId, newMache);
+        } catch (Exception e) {
+            LOG.error("failed adding map {} to cache", mapId, e);
+            throw new RuntimeException("Failed to create map", e);
+        }
+        return newMache;
+    }
+
+    /**
+     * Puts a key/value into the specified map, if the map does not exist it is created.
+     *
+     * @param mapId The map
+     * @param key   The key
+     * @param value The value
+     */
+    public void putKey(String mapId, String key, String value) {
+        Mache<String, String> mache = getMache(mapId);
+        mache.put(key, value);
+    }
+
+    /**
+     * Get the value for the given map and key.
+     *
+     * @param mapId The map
+     * @param key   The key
+     * @return The value
+     */
+    public String getKey(String mapId, String key) {
+        Mache<String, String> mache = getMache(mapId);
+        return mache.get(key);
+    }
+
+    private synchronized Mache<String, String> getMache(String mapId) {
+        Mache<String, String> mache = cacheInstances.get(mapId);
+        if (mache == null) {
+            mache = createMap(mapId);
+        }
+        return mache;
+    }
```

Comment:
Start the rest service against specific maps?

Let's pick this up tomorrow
On 27 Jan 2016 11:53, "JamieDrummond" <notifications@github.com> wrote:

> In
> mache-vertx/src/main/java/com/excelian/mache/vertx/MacheInstanceCache.java
> <https://github.com/Excelian/Mache/pull/45#discussion_r51021603>:
>
> > +     * @param mapId The map
> > +     * @param key   The key
> > +     * @return The value
> > +     */
> > +    public String getKey(String mapId, String key) {
> > +        Mache<String, String> mache = getMache(mapId);
> > +        return mache.get(key);
> > +    }
> > +
> > +    private synchronized Mache<String, String> getMache(String mapId) {
> > +        Mache<String, String> mache = cacheInstances.get(mapId);
> > +        if (mache == null) {
> > +            mache = createMap(mapId);
> > +        }
> > +        return mache;
> > +    }
>
> I think we may need to address this sooner as the Cassandra example now
> explicitly states a table to start against. Given the direction of recent
> changes it would seem the decision is to start the rest service against
> known maps and if requests are received outside of this configured scope
> then they are to be ignored.
>
> In the mentioned use case do the maps names remain constant overnight? If
> so we can perhaps have an endpoint allowing for eviction/reloading and a
> very simple cron job to evict at a specific time. This can already be
> achieved if the cron job sends a rabbit eviction message I believe?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Excelian/Mache/pull/45/files#r51021603>.
>